xblock/scorable.py

```diff
@@ -0,0 +1,152 @@
+
+        original_score = self.get_score()
+        if original_score is None:
+            self._publish_scoring_event('rescore_failure', {'failure': 'unanswered'})
+            raise ValueError(_('Problem must be answered before it can be rescored.'))
+
+        try:
+            new_score = self.calculate_score()
+        except Exception:  # pylint: disable=broad-except
+            self._publish_scoring_event('rescore_failure', {'failure': 'calculation error'})
+            return False
+
+        self.set_score(new_score)
+        self._publish_grade(only_if_higher)
+        self._publish_scoring_event(
+            'rescore_result',
+            {
+                'result': 'score updated',
+                'original_score': original_score._asdict(),  # pylint: disable=protected-access
+                'new_score': new_score._asdict(),  # pylint: disable=protected-access
+            }
+        )
+        return True
+
+    def allows_rescore(self):
+        """
+        Boolean value: Can this problem be rescored.
+        """
+        return True
+
+    def get_score(self):
+        """
+        Return a raw score already persisted on the XBlock.
+
+        Returns:
+            Score(earned=float, total=float)_
+        """
+        raise NotImplementedError
+
+    def set_score(self, score):
+        """
+        Persist a score to the XBlock.  The score is a named tuple with a score
+        attribute and a total, reflecting the raw earned score, and the maximum
+        raw score the student could have earned.
+
+        Arguments:
+            score: Score(earned=float, total=float)
+
+        Returns:
+            None
+        """
+        raise NotImplementedError
+
+    def calculate_score(self):
+        """
+        Calculate a new raw score based on the state of the problem.
+        This method should should not modify the state of the
+        XBlock.
+
+        Returns:
+            Score(earned=float, total=float)
+        """
+        raise NotImplementedError
+
+    def _base_event_info(self):
+        """
+        Return a dictionary containing information that belongs on any logged
+        events.
+        """
+        return {
+            'usage_key': six.text_type(self.location)
+        }
+
+    def _publish_grade(self, only_if_higher=None):
+        """
+        Publish a grade to the runtime.
+        """
+        score = self.get_score()
+        grade_dict = {
+            'value': score.earned,
+            'max_value': score.total,
+        }
+        if only_if_higher is not None:
+            grade_dict['only_if_higher'] = only_if_higher
+        self.runtime.publish('grade', grade_dict)
+
+    def _publish_scoring_event(self, name, data):
+        """
+        Publish an event related to scoring.
+        """
+        logged = {}  # Ensure we don't mutate our arguments
+        logged.update(data)
+        logged.update(self._base_event_info())
+        self.runtime.publish(name, logged)
```

Comment:
I think you might need to pass `self` as the first parameter to `self.runtime.publish` here and elsewhere. See https://github.com/edx/XBlock/blob/master/xblock/runtime.py#L503